### ISSUE TITLE
undefined reference to `__umodsi3'

### ISSUE THREAD

**Issue by ladmanj** [tkchia/gcc-ia16#89]
Hi

Linking of my project fails with undefined reference to `__umodsi3' and to `__udivsi3'

More detailed description of the project is here https://github.com/tkchia/gcc-ia16/issues/66#issuecomment-953421896

Please help.

```
ia16-elf-ld: mode elf_i386
attempt to open /home/ladmanj/DataDisk/src/build-ia16-master/prefix/lib/gcc/ia16-elf/6.3.0//libgcc.so failed
attempt to open /home/ladmanj/DataDisk/src/build-ia16-master/prefix/lib/gcc/ia16-elf/6.3.0//libgcc.a succeeded
/home/ladmanj/DataDisk/src/build-ia16-master/prefix/lib/gcc/ia16-elf/6.3.0//libgcc.a
attempt to open main.o succeeded
main.o
attempt to open memtest.o succeeded
memtest.o
attempt to open console.o succeeded
console.o
attempt to open arch_8086/io.o succeeded
arch_8086/io.o
attempt to open arch_8086/init.o succeeded
arch_8086/init.o
ia16-elf-ld: console.o: in function `xprintf':
(.text.xprintf+0x154): undefined reference to `__umodsi3'
ia16-elf-ld: (.text.xprintf+0x167): undefined reference to `__udivsi3'
ia16-elf-ld: link errors found, deleting executable `memtest.elf'
make: *** [Makefile:68: memtest.elf] Error 1
rm memtest.o console.o arch_8086/init.o main.o arch_8086/io.o
```

**Comment by tkchia** [comment]
Hello @ladmanj,

You need to place the libraries on the command line _after_ your own modules.  The way the linker works is, it needs to go through your modules in order to find out which functions/variables they need from the libraries.

The usual way to do this is to have an `$(LDLIBS)` make variable, which is separate from `$(LDFLAGS)`.  Then in your recipe, you can put `$(LDFLAGS)` before your own object files, but `$(LDLIBS)` after:

```
LDLIBS = -lgcc
...
%.elf: $(OBJECTS) memtest.ld
        $(LD) $(LDFLAGS) -o $@ $(OBJECTS) $(LDLIBS)
```

Let me know if there are any further issues.  Thank you!

**Comment by ladmanj** [comment]
Hi @tkchia 

Great suggestion, the symbols are now found successfully.

But it revealed a new problem unfortunately.

hidden symbol ` __ia16_ldivmodu ' isn't defined

```
==================================================
ia16-elf-ld: mode elf_i386
attempt to open main.o succeeded
main.o
attempt to open memtest.o succeeded
memtest.o
attempt to open console.o succeeded
console.o
attempt to open arch_8086/io.o succeeded
arch_8086/io.o
attempt to open arch_8086/init.o succeeded
arch_8086/init.o
attempt to open /home/ladmanj/DataDisk/src/build-ia16-master/prefix/lib/gcc/ia16-elf/6.3.0//libgcc.a succeeded
/home/ladmanj/DataDisk/src/build-ia16-master/prefix/lib/gcc/ia16-elf/6.3.0//libgcc.a
(/home/ladmanj/DataDisk/src/build-ia16-master/prefix/lib/gcc/ia16-elf/6.3.0//libgcc.a)udivsi3.o
(/home/ladmanj/DataDisk/src/build-ia16-master/prefix/lib/gcc/ia16-elf/6.3.0//libgcc.a)umodsi3.o
(/home/ladmanj/DataDisk/src/build-ia16-master/prefix/lib/gcc/ia16-elf/6.3.0//libgcc.a)ia16-ldivmodu.o
ia16-elf-ld: memtest.elf: hidden symbol ` __ia16_ldivmodu ' isn't defined
ia16-elf-ld: final link failed: bad value
make: *** [Makefile:69: memtest.elf] Chyba 1
rm memtest.o console.o arch_8086/init.o main.o arch_8086/io.o
```

**Comment by tkchia** [comment]
Hello @ladmanj,

Thanks for the report.  It turns out there was a long-standing bug (?) in GCC which was causing bogus references to symbols with leading and trailing spaces (`" __ia16_ldivmodu "` should be `"__ia16_ldivmodu"`, which _is_ defined).  Try to pull a more recent Git commit of `gcc-ia16` and try again.

Thank you!

**Comment by tkchia** [comment]
Hello @ladmanj,

Are you now able to link your code, with the new `gcc-ia16`?  Thank you!

**Comment by ladmanj** [comment]
Hi @tkchia 

Yes it helped.

Now i'm trying to access the whole memory space by reading and writing at __far pointer, but the address lines A18 and A19 are stuck low somehow :-(

Thank you

**Comment by tkchia** [comment]
Hello @ladmanj,

Good luck!  For the time being, you can still address 256 KiB of memory, I suppose. :-)

Thank you!

**Comment by ladmanj** [comment]
> Hello @ladmanj,
> 
> Good luck! For the time being, you can still address 256 KiB of memory, I suppose. :-)
> 
> Thank you!

Thank you. Of course 256KiB is plenty of room, but the rules of the game sounds differently.
I have specific hardware with no documentation. Most of the functionality is now known, but there is more to discover.

I have connected 1MiB RAM (instead of the original 8KiB, the RAM is overshadowed by the ROM during read access.
I have modified 0xffff0 code to copy whole ROM to the same location in  RAM and then disable the ROM.

I see, with my logic analyzer, that the copy is done, the code runs from RAM for a while but then it crashes (i have not enough LA lines to see all the address pins).

So i want to test the whole memory and the glue logic mainly to investigate why it doesn't work properly.

The code starts at 0xffff0 as usual, then the copy is done at 0xfffd0 and finaly it jumps to 0xf8000. The ROM, when connected, is mirrored to all 16 pages (\A15 is Chip Select and higher addresses are ignored)

This testing code jumps from 0xffff0 right to 0000:8000 and all the segment registers are set to 0000.
It seems now okay, but i need to sweep over all the "gaps" where the ROM is not selected and as it starts to work properly, disable the ROM and check the rest of the RAM.

Sorry for a long story to a closed topic.

Thank you for everything

JL
